rfd/0174-kube-rbac-bootstrap.md

```diff
@@ -0,0 +1,176 @@
+We will add an option to the teleport-kube-agent chart and to our
+kubeconfig generation script to not create those cluster role bindings, but it will be enabled by default. 
+In the teleport-kube-agent Helm chart we will add another field to the `rbac` section, `bindDefaultRoles`, that would control whether we expose the default
+roles through cluster bindings or not.
+
+Example content of a `cluster-values.yaml` file that could be used when installing the Helm chart:
+```yaml
+roles: kube,app,discovery
+authToken: test-auth-token
+proxyAddr: tele.local:3080
+kubeClusterName: test
+rbac:
+  create: true           <-- existing setting, defaults to true
+  bindDefaultRoles: true <-- new setting, also defaults to true
+labels:
+  teleport.internal/resource-id: b34f651c-32de-45f6-86dc-ab5173f716c9
+enterprise: true
+```
+
+If either `rbac.create` or `rbac.bindDefaultRoles` is false, we won't expose the default roles.
+
+### Automatically provisioning RBAC resources to Kubernetes clusters.
+
+We will add a new type of resources, called "KubeProvision", where user will be able to specify Kubernetes RBAC resources they want to
+be provisioned onto their Kubernetes clusters.
+clusters. A single Teleport resource of that type can define multiple Kubernetes RBAC resources. There's a limit on size of a resource Teleport
+can save on the backend, therefore, if some users would have unusually large amount of Kubernetes RBAC resources they want to provision, they would
+need to split it into two KubeProvision resources.
+
+```protobuf
+import "teleport/header/v1/metadata.proto";
+
+// KubeProvision represents a Kubernetes resources that can be provisioned on the Kubernetes clusters.
+// This includes roles/role bindings and cluster roles/cluster role bindings.
+// For rationale behind this type, see the RFD 174.
+message KubeProvision {
+  // The kind of resource represented.
+  string kind = 1;
+  // Not populated for this resource type.
+  string sub_kind = 2;
+  // The version of the resource being represented.
+  string version = 3;
+  // Common metadata that all resources share.
+  teleport.header.v1.Metadata metadata = 4;
+  // The specific properties of kube provision.
+  KubeProvisionSpec spec = 5;
+}
+
+// KubeProvisionSpec is the spec for the kube provision message.
+message KubeProvisionSpec {
+  // resources_data is base64 encoded YAML definitions of the Kubernetes resources.
+  string resources_data = 3;
+}
+```
+
+Every five minutes we will run a reconciliation loop that will compare the current state on Kubernetes clusters under
+Teleport management with the desired state and update the cluster's RBAC if there's a difference.
+
+Teleport will mark RBAC resources under its control with the "app.kubernetes.io/managed-by: Teleport" label. That way we will
+separate resources managed by Teleport and those managed by the user manually.
```

Comment:
Can we also set the ownerReference==teleport statefulset in the RBAC metadata? This will ensure cascading deletion.

Reply:
Turns out cluster scoped resources (like ClusterRole) can't have owner reference pointing to a namespaced resource.